### DISCUSSION TITLE
Strange behaviour when switching from rc.2 to rc.3 with tokio

### DISCUSSION THREAD

**Discussion by ogarcia** [rwf2/Rocket#2547] [Questions]
In the main function of [my project](https://gitlab.com/ogarcia/lesma/-/blob/master/src/main.rs#L65) I have declared [a process with tokio to run in the background](https://gitlab.com/ogarcia/lesma/-/blob/master/src/main.rs#L117). It has never given me any problem but after upgrading from rc.2 to rc.3 I get the following error.
```
thread 'main' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', src/main.rs:117:5
```

As I said I have never had a problem with this, the main function has always been declared as follows.
```rust
#[launch]
fn rocket() -> _ {
...
}
```

However, after updating the versions of tokyo and rocket to rc.3, I get the error I mentioned before and in order for it to work I have to declare the main function in this way (this is not yet reflected on GitLab).

```rust
#[launch]
#[tokio::main]
async fn rocket() -> _ {
...
}
```

Was this done intentionally or could it be a bug?

**Comment by SergioBenitez** [comment]
No, this shouldn't be necessary, and in fact doing that would mean two async runtimes are started. Are you sure that the only change you've made is upgrading to rc3? Could you try downgrading to rc2 (use `rocket = "=0.5.0-rc.2"` in your Cargo.toml then cargo update) and see if the problem really disappears?

There was no relevant change to the way Rocket starts the async runtime between rc2 and rc3, so I'm inclined to believe it isn't Rocket that's causing this behavior. ~~To confirm, could you share a reduced and ideally minimal erroring version of your code?~~ Apologies, I see you've linked the entire project. Once you confirm the above, I'll take a look at your code directly to see what's going on.

Edit: one possibility is that your project is pulling in multiple versions of tokio. Without doing anything else, could you check the output of `cargo tree | grep tokio` and see if you find two versions of tokio? If so, it would be great to know why Cargo has pulled two versions (check your resolver version, try a cargo update, etc.), but the easiest way to resolve the issue would be to remove your direct dependency on tokio and instead use it indirectly via Rocket's reexport, e.g. `rocket::tokio::spawn`.

**Reply by ogarcia** [reply]
I have checked the dependency tree and the versions are the same:
```
% cargo tree | grep tokio
│   │   ├── tokio v1.28.1
│   │   │   └── tokio-macros v2.1.0 (proc-macro)
│   │   └── tokio-util v0.7.4
│   │       ├── tokio v1.28.1 (*)
│   │   │   │   ├── tokio v1.28.1
│   │   │   ├── tokio v1.28.1 (*)
│   │   │   │   ├── tokio v1.28.1 (*)
│   │   │   │   ├── tokio-util v0.7.4 (*)
│   │   │   ├── tokio v1.28.1 (*)
│   │   ├── tokio v1.28.1 (*)
│   ├── tokio v1.28.1 (*)
│   ├── tokio-stream v0.1.11
│   │   └── tokio v1.28.1 (*)
│   ├── tokio-util v0.7.4 (*)
```

Anyway, I have done what you have told me and I have used the library that rocket re-exports and I still get the same error:
```
thread 'main' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', src/main.rs:117:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This is line 117 which as you can see is already used by the tokio from rocket.
```rust
    // Launch background cleaner
    rocket::tokio::spawn(async move {
        background::keep_house_clean(&storage).await;
    });
```

**Reply by ogarcia** [reply]
I just realised one thing, for it to work correctly I simply have to declare the `rocket()` function as async.

This works:
```rust
#[launch]
async fn rocket() -> _ {
...
}
```

But this not:
```rust
#[launch]
fn rocket() -> _ {
...
}
```

**Reply by kennycallado** [reply]
It works but breaks the tests.

**Reply by SergioBenitez** [reply]
> It works but breaks the tests.

Which tests?

**Reply by SergioBenitez** [reply]
> I just realised one thing, for it to work correctly I simply have to declare the `rocket()` function as async.
> 
> This works:
> 
> ```rust
> #[launch]
> async fn rocket() -> _ {
> ...
> }
> ```
> 
> But this not:
> 
> ```rust
> #[launch]
> fn rocket() -> _ {
> ...
> }
> ```

This is insightful, thank you for pointing this out.

When the function is not marked `async`, it's not run inside a runtime but merely executed to get the resulting instance of `Rocket`. This results in the error you're seeing: there is no runtime running. This is because, while Rocket _could_ always run the function inside of a runtime, doing so would result in the function operating differently when used as `main` and when used as a normal function. For example, calling the function from a unit test without the unit test being `async` would cause the function to fail when it would otherwise work just fine as `main`. We want the function to always fail or always succeed, irrespective of where it's run - this is what happens today.

I believe the best approach here is perhaps to emit a warning when something that requires a runtime is called in a non-`async` `launch`. Unfortunately, detecting that "something requires a runtime" is rather impossible, in general, but we could lint for `spawn` as a baseline.

**Reply by SergioBenitez** [reply]
Here's what I'll likely be pushing:

```rust
#[rocket::launch]
fn rocket() -> _ {
    use rocket::tokio;
    tokio::spawn(async move {

    });

    rocket::build()
}
```

```
warning: task is being spawned outside an async context
 --> src/main.rs:4:5
  |
4 | /     tokio::spawn(async move {
5 | |
6 | |     });
  | |______^
  |
help: declare this function as `async fn` to require async execution
 --> src/main.rs:2:1
  |
2 | fn rocket() -> _ {
  | ^^^^^^^^^^^^^^^^
note: `#[launch]` call is here
 --> src/main.rs:1:1
  |
1 | #[rocket::launch]
  | ^^^^^^^^^^^^^^^^^
```

Edit: this was pushed.
